### PR TITLE
[TypeScript] Utils

### PR DESCRIPTION
**Pull request by andrewrjohn**
#2

### CODE REVIEW

File: src/utils/base58.ts [hunk 1]
```diff
@@ -16,10 +16,7 @@ const LEADER = ALPHABET.charAt(0);
 const FACTOR = Math.log(BASE) / Math.log(256); // log(BASE) / log(256), rounded up
 const iFACTOR = Math.log(256) / Math.log(BASE); // log(256) / log(BASE), rounded up
 
-function encode(source) {
-  if (!Buffer.isBuffer(source)) {
-    throw new TypeError("Expected Buffer");
-  }
```
Comment (andrewrjohn) [inline src/utils/base58.ts:22]: Shouldn't need manual typechecks like this anymore because of TS 🎉

File: src/utils/base58.ts [hunk 2]
```diff
@@ -68,8 +65,7 @@ function encode(source) {
   return str;
 }
 
-function decode(source) {
-  if (typeof source !== "string") throw new TypeError("Expected String");
```
Comment (andrewrjohn) [inline src/utils/base58.ts:72]: Shouldn't need manual typechecks like this anymore because of TS 🎉

File: src/utils/reader_chunks.ts [hunk 3]
```diff
@@ -176,13 +176,15 @@ class BufferChunksReader {
...
+        throw new Error(
+          `Invalid length while reading varlength buffer. Expected to read: ${len} and read ${buf?.length}`
+        );
+      }
+      return buf;
     }
-    return buf;
   }
 
   readVarintBuf() {
```
Comment (andrewrjohn) [inline src/utils/reader_chunks.ts:190]: `this.buf` does not exist. @kevinejohn can you help here?

File: src/utils/reader_chunks.ts [hunk 4]
```diff
@@ -102,52 +102,52 @@ class BufferChunksReader {
 
   readUInt8() {
     const buf = this.read(1);
-    return buf.readUInt8();
+    return buf?.readUInt8();
```
Comment (andrewrjohn) [inline src/utils/reader_chunks.ts:105]: @kevinejohn maybe all of these read methods should throw an Error if `buf === undefined`. That way the return type of all these methods would be a defined type.

File: src/utils/reader.ts [hunk 5]
```diff
@@ -1,40 +1,39 @@
...
+  slice(i: number, j: number) {
     return this.buf.slice(i, j);
   }
 
-  set(obj) {
-    this.buf = obj.buf || this.buf || undefined;
-    this.pos = obj.pos || this.pos || 0;
-    return this;
-  }
-
```
[the remark this answers is not in the document]
Comment (kevinejohn) [reply src/utils/reader.ts:37]: no it wasn't. We can remove it

File: src/utils/reader_chunks.ts (hunk 3 above)
Comment (kevinejohn) [reply src/utils/reader_chunks.ts:190]: it just just be a local `const buf` and not `this.buf`. Am I missing something?

File: src/utils/reader_chunks.ts (hunk 4 above)
Comment (kevinejohn) [reply src/utils/reader_chunks.ts:105]: lets make it throw if it is `undefined`

**Pull request merged**